src/Workspaces/Core/Portable/Workspace/Solution/Project.cs

```diff
@@ -284,6 +283,34 @@ public async ValueTask<IEnumerable<SourceGeneratedDocument>> GetSourceGeneratedD
             return null;
         }
 
+        /// <summary>
+        /// Returns the <see cref="SourceGeneratedDocumentState"/> for a source generated document that has already been generated and observed.
+        /// </summary>
+        /// <remarks>
+        /// This is only safe to call if you already have seen the SyntaxTree or equivalent that indicates the document state has already been
+        /// generated. This method exists to implement <see cref="Solution.GetDocument(SyntaxTree?)"/> and is best avoided unless you're doing something
+        /// similarly tricky like that.
+        /// </remarks>
+        internal SourceGeneratedDocument? TryGetSourceGeneratedDocumentForAlreadyGeneratedId(DocumentId documentId)
+        {
+            // Easy case: do we already have the SourceGeneratedDocument created?
+            if (_idToSourceGeneratedDocumentMap.TryGetValue(documentId, out var document))
+            {
+                return document;
+            }
+
+            // Trickier case now: it's possible we generated this, but we don't actually have the SourceGeneratedDocument for it, so let's go
```

Comment:
wat?   if we generated this... how would we not have the SourceGeneratedDocument for it?

Reply:
So the generator would have ran, we have a SourceGeneratedDocument**State** available since that's the output of the compilation tracker. But we may have not yet made the wrapper type for it yet.